src/v/archival/ntp_archiver_service.cc

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/ntp_archiver_service.h"
+
+#include "archival/logger.h"
+#include "archival/utils.h"
+#include "model/metadata.h"
+#include "s3/client.h"
+#include "s3/error.h"
+#include "storage/disk_log_impl.h"
+#include "storage/fs_utils.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <fmt/format.h>
+
+#include <stdexcept>
+
+namespace archival {
+
+class arch_policy_iface {
+public:
+    virtual std::optional<manifest>
+    make_local_manifest(storage::log_manager& lm) = 0;
+};
+
+class arch_policy_non_compacted final : public arch_policy_iface {
+public:
+    arch_policy_non_compacted(model::ntp ntp, model::revision_id rev)
+      : _ntp(ntp)
+      , _rev(rev) {}
+
+    std::optional<manifest>
+    make_local_manifest(storage::log_manager& lm) final {
+        manifest tmp(_ntp, _rev);
+        std::optional<storage::log> log = lm.get(_ntp);
+        if (!log) {
+            return std::nullopt;
+        }
+        auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
+        if (plog == nullptr) {
+            return std::nullopt;
+        }
+        for (const auto& segment : plog->segments()) {
+            bool closed = !segment->has_appender(); // is_closed();
+            bool compacted = segment->is_compacted_segment();
+            if (!closed || compacted) {
+                continue;
+            }
+            manifest::segment_meta meta{
+              .is_compacted = segment->is_compacted_segment(),
+              .size_bytes = segment->size_bytes(),
+              .base_offset = segment->offsets().base_offset,
+              .committed_offset = segment->offsets().committed_offset,
+              .is_deleted_locally = false,
+            };
+            auto path = std::filesystem::path(segment->reader().filename());
+            auto seg_name = path.filename().string();
+            // generate segment path
+            tmp.add(segment_name(seg_name), meta);
```

Comment:
Interesting.... should we add a looping mechanics to the log concept. Wondering why we need to materialize all segments into memory structs. 

Question: do we release the memory in this metadata object at any other point?

Reply:
I wonder if we should do a similar thing, where we push one more `bit` into the segment and we can ask if the segment has been backed up ? 

Thoughts? 

cc: @dotnwat